systemctl: add varlink-based JSON output for manager properties

gvenugo3 opened the pull request:
## Summary
- Implement server-side JSON generation for `systemctl show` manager properties using varlink
- Add fallback to D-Bus when varlink socket is not available  
- Update man page documentation for the new functionality

## Changes
- **systemctl-show.c**: Add `show_manager_varlink_json()` function that uses `io.systemd.Manager.Describe` varlink method
- **systemctl.xml**: Document varlink-based JSON output behavior for manager properties

## Benefits
- Improved performance by avoiding client-side bus property to JSON conversion
- Server-side JSON generation as suggested in #31948
- Maintains backward compatibility with graceful fallback

## Test plan
- [x] Test `systemctl show --output=json` (should use varlink)
- [x] Test fallback when varlink socket unavailable
- [x] Verify existing functionality unchanged for non-JSON output
- [x] Test with and without unit arguments

poettering commented on src/systemctl/systemctl-show.c:
```diff
@@ -2415,6 +2417,24 @@ static int show_system_status(sd_bus *bus) {
+        _cleanup_(sd_json_variant_unrefp) sd_json_variant *reply = NULL;
+        int r;
+
+        r = sd_varlink_connect_address(&vl, "/run/systemd/io.systemd.Manager");
```
hmm, this is more complex: this needs to work if --machine= or --host= is used, i.e. connect to non-local services.

gvenugo3 replied:
Good catch! I've updated the implementation to handle remote connections properly.

  The varlink approach now only applies to local connections (arg_transport == BUS_TRANSPORT_LOCAL). For remote
  connections using --host= or --machine=, it falls back to the traditional D-Bus method.

  While varlink does technically support remote connections via SSH transport, systemctl doesn't currently integrate
  varlink with its --host/--machine infrastructure, so limiting to local connections is the correct approach for now.

  Updated the documentation as well to clarify this behavior.

yuwata replied:
No, this should not be 'fixed' by documentation, but use correct destination.

yuwata commented on src/systemctl/systemctl-show.c:
```diff
@@ -2489,6 +2491,24 @@ static int show_system_status(sd_bus *bus) {
+        r = sd_varlink_connect_address(&vl, "/run/systemd/io.systemd.Manager");
+        if (r < 0)
+                return log_error_errno(r, "JSON output for manager properties requires varlink support. "
+                                        "Varlink socket not available at /run/systemd/io.systemd.Manager: %m");
```
Please make the error message simpler, e.g. 
"Failed to connect to /run/systemd/io.systemd.Manager: %m"

yuwata commented on src/systemctl/systemctl-show.c:
```diff
@@ -2522,9 +2542,13 @@ int verb_show(int argc, char *argv[], void *userdata) {
-                                return show_one(bus, "/org/freedesktop/systemd1", NULL, show_mode, &new_line, &ellipsized);
+                                if (OUTPUT_MODE_IS_JSON(arg_output))
+                                        return show_manager_varlink_json();
+                                else
```
Unnecessary `else`. Please drop.

yuwata commented on man/systemctl.xml:
```diff
@@ -342,6 +342,12 @@ Jan 12 10:46:45 example.com bluetoothd[8900]: gatt-time-server: Input/output err
+            without specifying units), <command>systemctl</command> will use the varlink interface
+            <literal>io.systemd.Manager.Describe</literal> when available, which provides server-side JSON
+            generation for improved performance. If the varlink socket is not available, it falls back to
+            the traditional D-Bus method.</para>
```
As commented in the above, this is a regression caused by your PR. Please drop the comment.

yuwata replied:
This is very internal implementation, and not necessary to document in the man page. Please drop the entire paragraph.

yuwata commented on src/systemctl/systemctl-show.c:
```diff
@@ -2488,6 +2491,38 @@ static int show_system_status(sd_bus *bus) {
+
+        r = sd_varlink_call(vl, "io.systemd.Manager.Describe", NULL, &reply, NULL);
+        if (r < 0)
+                return log_error_errno(r, "Failed to call io.systemd.Manager.Describe: %m");
```
Please use varlink_call_and_log()

yuwata commented on src/systemctl/systemctl-show.c:
```diff
@@ -2488,6 +2491,38 @@ static int show_system_status(sd_bus *bus) {
+
+                address = path_join(runtime_dir, "systemd/io.systemd.Manager");
+                if (!address)
+                        return log_oom();
```
Please drop redundant oom check.

yuwata commented on src/systemctl/systemctl-show.c:
```diff
@@ -2488,6 +2491,38 @@ static int show_system_status(sd_bus *bus) {
+                        return log_oom();
+        } else
+                address = strdup("/run/systemd/io.systemd.Manager");
+
```
Please drop the unnecessary empty line.
We do not insert an empty line between function call and error check.

yuwata commented on src/systemctl/systemctl-show.c:
```diff
@@ -2521,9 +2556,13 @@ int verb_show(int argc, char *argv[], void *userdata) {
-                        if (show_mode == SYSTEMCTL_SHOW_PROPERTIES)
+                        if (show_mode == SYSTEMCTL_SHOW_PROPERTIES) {
                                 /* systemctl show --all → show properties of the manager */
+                                if (OUTPUT_MODE_IS_JSON(arg_output) && arg_transport == BUS_TRANSPORT_LOCAL)
```
Why not to use remote connection via ssh?

yuwata commented on src/systemctl/systemctl-show.c:
```diff
@@ -2521,9 +2556,13 @@ int verb_show(int argc, char *argv[], void *userdata) {
+                        if (show_mode == SYSTEMCTL_SHOW_PROPERTIES) {
                                 /* systemctl show --all → show properties of the manager */
+                                if (OUTPUT_MODE_IS_JSON(arg_output) && arg_transport == BUS_TRANSPORT_LOCAL)
+                                        return show_manager_varlink_json();
```
Even the man page says 'fall back to use dbus' but it does not fallback on any errors.

yuwata commented on man/systemctl.xml:
```diff
@@ -342,6 +342,7 @@ Jan 12 10:46:45 example.com bluetoothd[8900]: gatt-time-server: Input/output err
             <para>For details about many of these properties, see the documentation of the D-Bus interface
             backing these properties, see
             <citerefentry><refentrytitle>org.freedesktop.systemd1</refentrytitle><manvolnum>5</manvolnum></citerefentry>.</para>
+
```
Please drop the unnecessary insertion of the empty line.

yuwata commented on src/systemctl/systemctl-show.c:
```diff
@@ -51,6 +53,8 @@
 #include "systemctl-util.h"
 #include "terminal-util.h"
 #include "utf8.h"
+#include "varlink-io.systemd.Manager.h"
```
Hm? Unused?

yuwata commented on src/systemctl/systemctl-show.c:
```diff
@@ -2488,6 +2492,35 @@ static int show_system_status(sd_bus *bus) {
+                return r;
+
+        sd_json_variant_dump(reply, output_mode_to_json_format_flags(arg_output), stdout, NULL);
+        return 0;
```
Please propagate the failure:
```c
return sd_json_variant_dump(...);
```

yuwata commented on src/systemctl/systemctl-show.c:
```diff
@@ -2521,9 +2554,17 @@ int verb_show(int argc, char *argv[], void *userdata) {
+                                if (OUTPUT_MODE_IS_JSON(arg_output)) {
+                                        r = show_manager_varlink_json();
+                                        if (r >= 0)
+                                                return r;
```
The returned value is unused. So,
```c
if (show_manager_varlink_json() >= 0)
        return 0;
```

yuwata commented on src/systemctl/systemctl-show.c:
```diff
@@ -2521,9 +2554,17 @@ int verb_show(int argc, char *argv[], void *userdata) {
-                        if (show_mode == SYSTEMCTL_SHOW_PROPERTIES)
+                        if (show_mode == SYSTEMCTL_SHOW_PROPERTIES) {
                                 /* systemctl show --all → show properties of the manager */
+                                if (OUTPUT_MODE_IS_JSON(arg_output)) {
```
Not only the json output case, we should always use varlink if exists.

yuwata reviewed (changes requested):
Still `--machine=` is not supported...

poettering commented on src/systemctl/systemctl-show.c:
```diff
@@ -2488,6 +2491,58 @@ static int show_system_status(sd_bus *bus) {
+                if (!runtime_dir)
+                        return log_error_errno(SYNTHETIC_ERRNO(EUNATCH), "XDG_RUNTIME_DIR not set.");
+
+                socket_path = strjoina(runtime_dir, "/systemd/io.systemd.Manager");
```
please use regular strjoin(), given that `runtime_dir` is user controlled, and might be used to voerflow the stack

gvenugo3 committed:
systemctl: add varlink-based JSON output for manager properties

Use the io.systemd.Manager.Describe varlink method for outputting
manager properties when available, falling back to D-Bus when the
varlink connection fails. This implements server-side JSON generation,
avoiding client-side bus property to JSON conversion.

Transport support:
- LOCAL/CAPSULE: connect directly to the varlink socket
- REMOTE (--host=): connect via SSH using ssh: URL scheme
- MACHINE (--machine=): falls back to D-Bus (container support via
  exec: URL scheme with nsenter is left as future work)

The socket path is selected based on runtime scope:
- System: /run/systemd/io.systemd.Manager
- User: $XDG_RUNTIME_DIR/systemd/io.systemd.Manager

Closes #31948

yuwata commented on src/systemctl/systemctl-show.c:
```diff
@@ -2515,9 +2574,13 @@ int verb_show(int argc, char *argv[], void *userdata) {
-                        if (show_mode == SYSTEMCTL_SHOW_PROPERTIES)
+                        if (show_mode == SYSTEMCTL_SHOW_PROPERTIES) {
                                 /* systemctl show --all → show properties of the manager */
+                                if (show_manager_varlink_json() >= 0)
```
This unconditionally shows the result in JSON format???

keszybz commented on src/systemctl/systemctl-show.c:
```diff
@@ -2482,6 +2484,62 @@ static int show_system_status(sd_bus *bus) {
+        int r;
+
+        /* Determine the socket path based on scope */
+        if (arg_runtime_scope == RUNTIME_SCOPE_USER) {
+                const char *runtime_dir = secure_getenv("XDG_RUNTIME_DIR");
+                if (!runtime_dir)
+                        return log_error_errno(SYNTHETIC_ERRNO(EUNATCH), "XDG_RUNTIME_DIR not set.");
+
+                socket_path = strjoin(runtime_dir, "/systemd/io.systemd.Manager");
+                if (!socket_path)
+                        return log_oom();
+        } else {
+                socket_path = strdup("/run/systemd/io.systemd.Manager");
+                if (!socket_path)
+                        return log_oom();
+        }
+
+        /* Build the appropriate varlink URL based on transport type */
+        switch (arg_transport) {
+        case BUS_TRANSPORT_LOCAL:
+        case BUS_TRANSPORT_CAPSULE:
+                address = strdup(socket_path);
+                break;
+
+        case BUS_TRANSPORT_REMOTE:
+                if (!arg_host)
+                        return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "Remote host not specified.");
+
+                address = strjoin("ssh:", arg_host, ":", socket_path);
+                break;
+
+        case BUS_TRANSPORT_MACHINE:
+                /* TODO: Implement container support via exec: URL scheme with nsenter */
+                return log_debug_errno(SYNTHETIC_ERRNO(ENOTSUP), "Varlink not yet supported for --machine=, falling back to D-Bus.");
+
+        default:
+                return log_error_errno(SYNTHETIC_ERRNO(EPROTONOSUPPORT), "Unsupported transport type.");
+        }
+        if (!address)
+                return log_oom();
```
Please this whole part out to a helper function, `manager_acquire_varlink_address` or smth.

keszybz commented on src/systemctl/systemctl-show.c:
```diff
@@ -2482,6 +2484,62 @@ static int show_system_status(sd_bus *bus) {
+
+        case BUS_TRANSPORT_MACHINE:
+                /* TODO: Implement container support via exec: URL scheme with nsenter */
+                return log_debug_errno(SYNTHETIC_ERRNO(ENOTSUP), "Varlink not yet supported for --machine=, falling back to D-Bus.");
```
We always use EOPNOTSUPP.

keszybz commented on src/systemctl/systemctl-show.c:
```diff
@@ -2482,6 +2484,62 @@ static int show_system_status(sd_bus *bus) {
+        switch (arg_transport) {
+        case BUS_TRANSPORT_LOCAL:
+        case BUS_TRANSPORT_CAPSULE:
+                address = strdup(socket_path);
```
just `TAKE_PTR(socket_path)`, no need to duplicate.

EDIT: I don't think this works at all. On my system I see:
`Invalid URL '/run/systemd/io.systemd.Manager': does not contain a ':'`

keszybz reviewed (changes requested):
So… this works in the limited circumstances, i.e. the showing of the state of the manager. But at the same time, this is just a tiny step. Ideally, we'd have the whole systemctl unit introspection state done over Varlink.

Dunno, it's not harmful to merge this as a first step (if the shortcomings are fixed), but at the same time, it'd be much better to do the whole thing.

Atemu commented on src/systemctl/systemctl-show.c:
```diff
@@ -2515,9 +2573,13 @@ int verb_show(int argc, char *argv[], void *userdata) {
-                        if (show_mode == SYSTEMCTL_SHOW_PROPERTIES)
+                        if (show_mode == SYSTEMCTL_SHOW_PROPERTIES) {
                                 /* systemctl show --all → show properties of the manager */
+                                if (OUTPUT_MODE_IS_JSON(arg_output) && show_manager_varlink_json() >= 0)
```
This falls back to showing the output in the bespoke systemd serialisation format. I find that to be unexpected when the user explicitly requested the output format to be json. If it's unable to produce json, you'd want the command producing json output to fail before the output reaches the JSON parser on the other end.

Mirroring https://github.com/systemd/systemd/pull/31948#pullrequestreview-1959482285, this should probably also be a separate `--json` flag.

(SN: the `--output` flag is quite a misnomer tbh.; it's not clear or intuitive that this only concerns the output format of journal entries. `--journal-format` or `--log-format` would be more obvious.)
